### ISSUE TITLE
Request

### ISSUE THREAD

**Issue by anotatta** [rk700/PyMuPDF#128]
Please add wrapper functions to set boxes (mediabox, cropbox, etc) and also to get the bounding box for Text, Image, and Path.  Thanks.

**Comment by JorjMcKie** [comment]
* `setCropBox` - underway, no problem.
* `setMediaBox` - while technically no problem, this shouldn't be done, because it invalidates the `/Contents` commands of the page.
* `MediaBox`, `CropBox` - properties, ok
* `BBox` - a function to get the bbox of an entry in `Page.getTextWords()` / `Page.getTextBlocks()`: **this what you mean?** 
* ``BBox`` of a **"path"**? What do you mean? Please be more specific, thanks

**Comment by anotatta** [comment]
Ok, I 'll be content with setcropbox method and mediabox&cropbox properties.

For bounding box, perhaps something equivalent to: fz_bound_path, fz_bound_text, fz_bound_annot.
I suppose it could be something like Page.getTextBound(), etc ...
The older version of mupdf has function to get bounding box for image also. 

Thank you for this great project.

**Comment by JorjMcKie** [comment]
Great to hear you PyMuPDF useful!
I believe we already have most of the bbox stuff (apart from this cropbox / mediabox business).

* We have `annot.rect`, `annot.setRect`
* We have `displaylist.rect`
* We have `page.rect`
* We do not support **`fz_shade`** - They, too, have a bbox. I never looked into it.
* We do not support **`fz_path`** nor **`fz_text`** nor do we plan to - **but** look further down to what we have!
    * `fitz.Rect(block[:4])` where `block `is an entry of `page.getTextBlocks(images = True)` and could be both, a text block or an image block. Of course one could wrap this under the name `bbox(block)` but I don't see much gain here. Esp. because such a function would not be bound to an object.
    * We have the `Shape` class which you do not find in MuPDF. It's pure Python string manipulation basically. It is used as a buffer for commands of the page's `/Contents` object. Among the supported commands are line, rect, curve, pie segments, circles, elipses, polyline, zigza and squiggly lines. I guess this is close to **`fz_path`**.
    * Drawings in a `Shape `are grouped via its `finish `method. It applies common attributes to the preceeding drawings (colors, dashing, ...).
    * We can insert **text boxes** into a `Shape`, which you may view as sort of **`fz_text`**. They already come with a bbox as part of their creation.

So in summary, the only real missing bbox is one that would contain drawings of a group (as defined by a `finish`) or the whole shape.

Before I do anything I would request your help here - at least for the high level design. Please have a look in the docs for your answers:
* I could add a `rect` property to the `Shape` class, initialized with `None`. Draw commands would modify it, but only if they go outside. Should it be reset at any time? As part of **finish**? As part of `commit`? Never?
* In `Shape` you will find a facility which I called "morphing": an argument for `finish` and `insertTextbox`. Can be used to distort the stuff almost arbitrarily. This facility remains agnostic of the new properties, right? wrong?

I concede, that a rect of a Shape can be useful when you want to appear a drawing as filling the whole page via setting its cropbox to it.

**Comment by JorjMcKie** [comment]
More details for my refusal adding a setMediaBox:
* You can add a new page, obviously always defining its mediabox as part of the process. The dimensions are under your control.
* If a page is not empty, then changing its mediabox will make it invalid - guarantied.
* If a page is indeed empty, you can delete it and insert a new page into the slot with desired dimensions.

**Comment by anotatta** [comment]
Sorry, I missed the **Shape** class -- I was looking at the older version (1.10) of the pymupdf manual.

I read the manual for 1.12.  Indeed **Annot.Rect()** and **Page.getTextBlocks(images=false)** provide bbox.  While the bbox of **Annot** and **TextBlocks** is readily accessible, **Page.getPageImageList()** provides the width & height of the image, without complete bbox or coordinate of where the image is located in the page.  I suppose the bbox for an image can be somehow obtained by using **displaylist.rect** , yes?  I don't know how to use this low level function yet, isn't this like drawing out the image to find out it's bbox?

Regarding your questions about **Shape** class.
1.  **rect** properties in **Shape** class
    For the most part the **bbox/rect** is meaningfull in relationship to the page. There might be a case where a user want to force a shape to be contained only within the user specified bbox.
    So, **rect** maybe provided after **finish**, certainly after **committ**, and never **never**.	
	
2.  If users want to morph any **Shape**, they should watch out what they are [**morphing**](http://www.imdb.com/title/tt0091064/) into.  They may provide a bbox where the morphed shaped is contained.  If not then upon **committ** the **rect** should be available.  If upon **committ** the **rect** is not provided, I guess the work around is to save the file, reopen it, and call for example **Page.getTextBlocks(images=false)** again to get the **rect**.  Really infefficient.

In conclusion most of the enhancement I requested are available already, except setCropBox. Time to close it?

**Comment by JorjMcKie** [comment]
Thanks for your input. Let me think about it a bit, before I come back - already with some experience, as I am already experimenting with a `Shape.rect`.

**Comment by JorjMcKie** [comment]
As part of the recent hotfixes to issues #130, #132, I have included an additional attribute `Shape.rect` as discussed here. Documentation is updated accordingly.

In a nutshell, it works like this:

- After  `Shape` object creation, and after `Shape.commit`, `Shape.rect` is `None`.
- `Shape` methods maintain `Shape.rect` - i.e. they always enlarge the rectangle, so that it encloses what has been drawn so far. Morphings are ignored.
- The programmer can always modify this attribute. E.g. he can morph the rectangle when and where required.
- Expected typical use is setting `Page.CropBox = Shape.rect` with the rectangle value taken immediately before `Shape.commit`.

I am going to close this issue for now - awaiting your feedback.
